### PR TITLE
Chapter 10 Exercise 02

### PR DESCRIPTION
**Pull request by archiebrowne**

### CODE REVIEW

**Comment by kbuzzard** [comment]
Can you fix the conflicts?

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 1]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+  · rw [hp, ← neg_lt_iff_pos_add']
+    calc
+      -s = (b - s) - b := by norm_num
+      _ < (b - s) / b * b := by exact Int.sub_lt_div_mul_self hb
+  · linear_combination hst
 
 /-
-Part b asks us to find such s and t for the examples in exercise 01, Lean provides us with these: 
+helper_2 builds on helper_1 but now adds the restriction that the new coefficient t' satisfies t' < 0.
+This effectivley solves the question, save for some changes of sign and the assumption that we can
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]:
```suggestion
This effectively solves the question, save for some changes of sign and the assumption that we can 
```

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 2]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+helper_2 builds on helper_1 but now adds the restriction that the new coefficient t' satisfies t' < 0.
+This effectivley solves the question, save for some changes of sign and the assumption that we can 
+initially find any s, t such that s * a + t * b = gcd(a, b).
+-/
+
+lemma helper_2 (a b n : ℤ) (ha : 0 < a) (hb : 0 < b) (hab : ∃ (s t : ℤ), s * a + t * b = n) : 
+    ∃ (s' t' : ℤ), 0 < s' ∧ t' < 0 ∧ s' * a + t' * b = n := by 
+    match helper_1 a b n hb hab with
+    |⟨s, t, ⟨h1, h2⟩⟩ => 
+    apply Or.elim (Classical.em (t < 0))
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]: This is usually not the right way to do do a case split: note that in the second goal you get `\not (t < 0)` which needs to be massaged with `push_neg`. Why not just do `rcases lt_or_le t 0 with (ht | ht)`? This achieves the same thing and it's cleaner.

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 3]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+    <;> intro ht
+    · use s, t
+      repeat (any_goals constructor)
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]: I get linter warnings telling me that there aren't any goals left at this point. Don't you?

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 4]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+      all_goals assumption
+    · push_neg at ht
+      set p := (t + a) / a with hp
+      have hpos : 0 ≤ p := by 
+        · rw [hp]
+          refine Iff.mpr (Int.le_ediv_iff_mul_le ha) ?_
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]: I think that's just a complicated way of doing `rw [Int.le_ediv_iff_mul_le ha]` (which could be added to the previous rewrite).

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 5]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+          norm_num
+          exact Int.add_nonneg ht (show 0 ≤ a by exact Int.nonneg_of_pos ha)
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]:
```suggestion
          exact Int.add_nonneg ht ha.le
```
Lemmas like a < b => a <= b can all be applied using dot notation.

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 6]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+      use (s + p * b), (t - p * a)
+      repeat (any_goals constructor)
+      · exact add_pos_of_pos_of_nonneg h1 (show _ by exact Iff.mpr (zero_le_mul_right hb) hpos)
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]:
```suggestion
      · exact add_pos_of_pos_of_nonneg h1 (Iff.mpr (zero_le_mul_right hb) hpos)
```
`show _ by exact` doesn't do anything here.

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 7]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+      · rw [sub_neg, hp]
+        have := @Int.sub_lt_div_mul_self (t + a) a ha
+        simp at this
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]: This is a non-terminal `simp`. Just use `simpa`.

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 8]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+The final statment of the question consistient with that in the book. helper_2 states
+the equality s' * a + t' * b = n with 0 < s' and t' < 0. 
+
+However, we want s' * a - t' * b = n with 0 < s' and 0 < t'. These are of course equivalent,
+and part_a deals with this. 
+-/
+
+lemma part_a (a b : ℤ) (ha : 0 < a) (hb : 0 < b) : 
+  ∃ (s t : ℤ), 0 < s ∧ 0 < t ∧ Int.gcd a b = s * a - t * b := by 
+  have := helper_2 a b (Int.gcd a b) ha hb (show _ by
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]:
```suggestion
  have := helper_2 a b (Int.gcd a b) ha hb (by
```

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 9]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+
+However, we want s' * a - t' * b = n with 0 < s' and 0 < t'. These are of course equivalent,
+and part_a deals with this. 
+-/
+
+lemma part_a (a b : ℤ) (ha : 0 < a) (hb : 0 < b) : 
+  ∃ (s t : ℤ), 0 < s ∧ 0 < t ∧ Int.gcd a b = s * a - t * b := by 
+  have := helper_2 a b (Int.gcd a b) ha hb (show _ by 
+    use Int.gcdA a b, Int.gcdB a b
+    simp [mul_comm]
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]: This is a non-terminal `simp`. Just close the goal with `simp [mul_comm, Int.gcd_eq_gcd_ab])`

File: M1FExplained/Chapter10/Exercises/Exercise02.lean [hunk 10]
```diff
@@ -2,14 +2,117 @@ import Mathlib.Tactic
...
+    exact Eq.symm (Int.gcd_eq_gcd_ab a b))
+  match this with
+  |⟨s, t, ⟨hs, ht, hst⟩⟩ => 
+  use s, -t
+  repeat (any_goals constructor)
+  · assumption
+  · exact Int.neg_pos_of_neg ht
+  · rw [Int.neg_mul]
+    simp
```
Comment (kbuzzard) [inline M1FExplained/Chapter10/Exercises/Exercise02.lean]: Similarly here (and the `id` in the next line isn't doing anything either). Just close the goal with `simp [hst]`

**Comment by archiebrowne** [comment]
changes pushed

**Comment by kbuzzard** [comment]
Thanks!

**Pull request merged**